usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+        {CKA_UNWRAP,            &unwrap,            sizeof(unwrap)},
+        {CKA_SIGN,              &sign,              sizeof(sign)},
+        {CKA_SIGN_RECOVER,      &sign_recover,      sizeof(sign_recover)},
+        {CKA_VERIFY,            &verify,            sizeof(verify)},
+        {CKA_VERIFY_RECOVER,    &verify_recover,    sizeof(verify_recover)},
+        {CKA_DERIVE,            &derive,            sizeof(derive)},
+        {CKA_EXTRACTABLE,       &extractable,       sizeof(extractable)},
+        {CKA_LOCAL,             &local,             sizeof(local)},
+        {CKA_NEVER_EXTRACTABLE, &never_extractable, sizeof(extractable)},
+        {CKA_MODIFIABLE,        &modifiable,        sizeof(modifiable)},
+        {CKA_WRAP_WITH_TRUSTED, &wrap_with_trusted, sizeof(wrap_with_trusted)},
+        {CKA_KEY_TYPE,          &template_keytype,  sizeof(template_keytype)},
+        {CKA_VALUE_LEN,         &valuelen,          sizeof(valuelen)},
+    };
+    CK_ULONG i;
+    CK_ATTRIBUTE *attr;
+    CK_BBOOL cktrue = TRUE;
+
+    rc = dll_m_GetAttributeValue(blob, blob_len, attrs,
+                                 sizeof(attrs) / sizeof(CK_ATTRIBUTE),
+                                 ep11_data->target);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Retrieving attributes from AB unwrapped key failed, rc=0x%lx\n",
+                    rc);
+        return rc;
+    }
+    if (template_keytype != keytype) {
+        TRACE_ERROR("Template specifies different key type than the AB key (0x%08lx vs 0x%08lx)\n",
+                    template_keytype, keytype);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+    for (i = 0; i < sizeof(attrs) / sizeof(CK_ATTRIBUTE); ++i) {
+        rc = build_attribute(attrs[i].type, attrs[i].pValue,
+                             attrs[i].ulValueLen, &attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("failed to build attribute (rc=0x%lx)\n", rc);
+            return rc;
+        }
+        rc = template_update_attribute(obj->template, attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
+                        __func__, rc);
+            return rc;
+        }
+    }
+    /* force sensitive */
+    rc = build_attribute(CKA_SENSITIVE, &cktrue, sizeof(CK_BBOOL), &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("failed to build attribute (rc=0x%lx)\n", rc);
+        return rc;
+    }    
+    rc = template_update_attribute(obj->template, attr);
+    if (rc != CKR_OK)
+        TRACE_ERROR("%s failed to force template sensitive; rc=0x%08lx\n", __func__, rc);
+    return rc;
+} 
+
+static CK_RV force_ab_sensitive(CK_ATTRIBUTE_PTR *p_attrs, CK_ULONG *p_attrs_len,
+                                CK_KEY_TYPE type)
+{
+    CK_RV rc;
+    CK_BBOOL cktrue = TRUE, abvalue, sensitive;
+
+    rc = get_bool_attribute_by_type(*p_attrs, *p_attrs_len,
+                                    CKA_IBM_ATTRBOUND, &abvalue);
+    if (rc == CKR_OK && abvalue) {
```

Comment:
We would save some brackets and indentation if we check for (rc != CKR_OK || !abvalue) and exit/goto end in this case.

Reply:
Yes, but I do not see a benefit here.  I find conjunctions easier than disjunctions and since the code is equivalent I would prefer to not change it.